Adds docs on 404 page

apreshill opened the pull request:
Closes https://github.com/rstudio/bookdown/issues/1208

@cderv I tested all the possibilities out for placement that you listed, and ultimately thought it would be (possibly) best placed as a "component" (https://bookdown.org/yihui/bookdown/web-pages-and-shiny-apps.html) since some of those at the bottom of the list were limited to LaTeX or HTML output. I'm agnostic on this placement though- it can work in any of the places you listed too. LMK what you think.

cderv commented on inst/examples/02-components.Rmd:
```diff
@@ -850,3 +850,15 @@ knitr::include_app('https://yihui.shinyapps.io/miniUI/', height = '600px')
 Shiny apps may take even longer to load than usual URLs. You may want to use a conservative value for the `delay` option, e.g., 10. Needless to say, `include_url()` and `include_app()` require a working Internet connection, unless you have previously cached the chunk (but web pages inside iframes still will not work without an Internet connection).
+
+## HTML 404 pages {#html-404}
+
```
I like the fact that the 404 section in Quarto doc on this page (https://quarto.org/docs/websites/website-navigation.html)
mentions two things: 

* A brief mention about 404 page are for 404 error with a link to know more. 
* Mention that service need to support it (Netlify, Github page)

I'll add sugestion below based on that.

cderv commented on inst/examples/02-components.Rmd:
```diff
@@ -850,3 +850,15 @@ knitr::include_app('https://yihui.shinyapps.io/miniUI/', height = '600px')
+
+## HTML 404 pages {#html-404}
+
+By default, both the `gitbook()` (Section \@ref(gitbook-style)) and `bs4_book()` (Section \@ref(bs4-book)) HTML output formats include a 404 page for readers who try to access a page in your book that cannot be found. The default 404.html is created using simple content (a header, and a body of 2 paragraphs). 
```
```suggestion
If a readers try to try to access a page in your book that cannot be found (e.g a link to a page that no more exists), a browser will display a [404 error](https://en.wikipedia.org/wiki/HTTP_404) as it can’t find the requested web page. This 404 error is displayed on a 404 page that each browser has a default for. You may want to customized this page to be in the style of your HTML book. 
Most web serving platforms (e.g. Netlify, Github Pages, or Gitlab Pages) will use a file named `404.html` in the root of your website as a custom error page if you provide it. 

With HTML book format, **bookdown** creates a default `404.html`  using simple content (a header, and a body of 2 paragraphs).  
```

cderv commented on inst/examples/02-components.Rmd:
````diff
@@ -850,3 +850,15 @@ knitr::include_app('https://yihui.shinyapps.io/miniUI/', height = '600px')
+knitr::include_graphics('images/404.png')
+```
+
+To customize the 404 page instead of using the default, you may add either a `_404.Rmd` or a `_404.md` file to your project root. If either file is found when you render the book, the content will be rendered as an `html_fragment` and will be included as the body of the 404 page. This means that the overall structure of your book's website (including navbar, footer, sidebars, etc.) and the CSS st ...
````
```suggestion
To customize the 404 page instead of using the default, you may add either a `_404.Rmd` or a `_404.md` file to your project root. If either file is found when you render the book, the content will be rendered as an `html_fragment()` and will be included as the body of the 404 page. This means that the overall structure of your book's website (including navbar, footer, sidebars, etc.) and the CSS styling will be preserved on the 404 page.
```

cderv commented on inst/examples/02-components.Rmd:
```diff
@@ -850,3 +850,15 @@ knitr::include_app('https://yihui.shinyapps.io/miniUI/', height = '600px')
+
+To customize the 404 page instead of using the default, you may add either a `_404.Rmd` or a `_404.md` file to your project root. If either file is found when you render the book, the content will be rendered as an `html_fragment` and will be included as the body of the 404 page. This means that the overall structure of your book's website (including navbar, footer, sidebars, etc.) and the CSS st ...
+
+If a 404.html file already exists in the main repo at the root level with the `.Rmd` files, then **bookdown** will leave that file as is and will not overwrite it.
```
I want to add this to explain
```suggestion
If a 404.html file already exists in the main repo at the root level with the `.Rmd` files, then **bookdown** will leave that file as is and will not overwrite it. This is because we assume you already have a mechanism in place in your publishing workflow to use this custom `404.html` you had setup.
```

but maybe unnecessary

cderv reviewed (commented):
In fact, I like where you put that. Having it in the component makes sense. 

I added a few comments mainly to add some precision that I think are missing. If you don't find that necessary, you can ignore. 

Thanks a lot !

apreshill commented:
Thanks for the feedback- I think this PR is all sets!

cderv reviewed (approved):
LGTM!

yihui commented on inst/examples/02-components.Rmd:
```diff
@@ -850,3 +850,19 @@ knitr::include_app('https://yihui.shinyapps.io/miniUI/', height = '600px')
+
+## HTML 404 pages {#html-404}
+
+If a reader tries to access a page in your book that cannot be found, a browser will display a [404 error](https://en.wikipedia.org/wiki/HTTP_404) as it cannot find the requested web page. This 404 error is displayed on a 404 page. Each browser has a default for a 404 page. However, most web serving platforms like Netlify, Github Pages, and Gitlab Pages will use a file named `404.html` in the roo ...
```
The default 404 page comes from the server, not browser:

```suggestion
If a reader tries to access a page in your book that cannot be found, a browser will display a [404 error](https://en.wikipedia.org/wiki/HTTP_404) as it cannot find the requested web page. This 404 error is displayed on a 404 page. Each web server has a default for a 404 page. However, most web serving platforms like Netlify, Github Pages, and Gitlab Pages will use a file named `404.html` in the root of your website as a custom error page, if you provide it. 
```

cderv replied:
Yes server usually have défaut 404. It is more common. I am quite certain that browser can also have there default 404 pages, but I guess it could have been mainly IE on windows.

yihui commented on inst/examples/02-components.Rmd:
```diff
@@ -850,3 +850,19 @@ knitr::include_app('https://yihui.shinyapps.io/miniUI/', height = '600px')
+
+As you can see, this 404 page is embedded within the book so that readers can quickly find their way back to the book's content. The overall structure of the book's website (including navbar, footer, sidebars, etc.) and the CSS styling are preserved on the 404 page.
+
+To customize the 404 page instead of using the one **bookdown** provides, you may add either a `_404.Rmd` or a `_404.md` file to your project root. If either file is found when you render the book, the content will be rendered as an `html_fragment()` and will be included as the body of the 404 page embedded within the book structure.
```
I suggest we omit the implementation detail `html_fragment()`:

```suggestion
To customize the 404 page instead of using the one **bookdown** provides, you may add either a `_404.Rmd` or a `_404.md` file to your project root. If either file is found when you render the book, the content will be rendered and included as the body of the 404 page embedded within the book structure.
```

yihui reviewed (approved):
Thanks! Just two minor suggestions.

The screenshot can be minified by `xfun::tinify()` (you'll need an API key from https://tinypng.com/developers). Below is the tinified version that I got, and you can use it to replace the current version in the PR:

![404](https://user-images.githubusercontent.com/163582/127568454-1a240608-be8e-4209-a054-ffe384516b9a.png)

BTW, the trickiest thing when including images is to make sure it looks fine in PDF output, which means you may need to set a proper `output.width` like `"80%"`. It may be possible to automatically calculate a proper percentage, but I have never found a chance to try.

cderv commented:
> The screenshot can be minified by `xfun::tinify()`

Is this the new way ? 

We were discussing optimization with `xfun::optipng()` that we are trying to do each time. Should we change tooling for this ?

yihui commented:
`optipng()` is lossless compression; `tinify()` is lossy but the quality loss is barely noticeable. `tinify()` usually reduces the file size much more significantly. I usually call `optipng()` first and pass the image to `tinify()` (as I'm typing here, I feel I should do this automatically inside `tinify()`).

cderv commented:
Oh I see so you do both. Thanks for the precision !

apreshill commented:
Move to chapter 6, see https://github.com/rstudio/bookdown/issues/1208#issuecomment-889482511

apreshill commented:
OK I moved to chapter 6, but wanted to add a section on metadata there too since the 404 page section felt odd to be on its own. This means I added 2 more png screenshots, but I cannot get a tinify API key 
<img width="678" alt="Screen Shot 2021-07-30 at 1 45 35 PM" src="https://user-images.githubusercontent.com/12160301/127709925-c71b9fc9-b6ff-4581-8ba8-66707eee1085.png">

If someone would help me with the images, I would appreciate it.

yihui commented:
I'll take care of the images. 

> add a section on metadata there too since the 404 page section felt odd to be on its own

Good idea. I wonder if it would be better to be the first or last section in this chapter (asking since these features are not essential).

> I cannot get a tinify API key

I can submit your rstudio email address on that page if you don't mind.

apreshill commented:
@yihui yes thank you if you can take care of the images. I moved the section lower in the chapter.
